llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp

```diff
@@ -4516,6 +4516,35 @@ SDValue TargetLowering::SimplifySetCC(EVT VT, SDValue N0, SDValue N1,
         }
       }
     }
+
+    // Optimize
+    //    (setcc (shift N00, N01C), 0, eq/ne) -> (setcc N00, 0, eq/ne)
+    // If all shifted out bits are known to be zero, then the zero'd ness
+    // doesn't change and we can omit the shift.
+    // If all shifted out bits are equal to at least one bit that isn't
+    // shifted out, then the zero'd ness doesn't change and we can omit the
+    // shift.
+    if ((Cond == ISD::SETEQ || Cond == ISD::SETNE) && C1.isZero() &&
+        N0.hasOneUse() &&
+        (N0.getOpcode() == ISD::SHL || N0.getOpcode() == ISD::SRL ||
+         N0.getOpcode() == ISD::SRA)) {
+      bool IsRightShift = N0.getOpcode() != ISD::SHL;
+      SDValue N00 = N0.getOperand(0);
+      // Quick checks based on exact/nuw/nsw flags.
+      if (IsRightShift ? N0->getFlags().hasExact()
+                       : (N0->getFlags().hasNoUnsignedWrap() ||
+                          N0->getFlags().hasNoSignedWrap()))
+        return DAG.getSetCC(dl, VT, N00, N1, Cond);
+      // More expensive checks based on known bits.
+      if (const APInt *ShAmt = DAG.getValidMaximumShiftAmountConstant(N0)) {
+        KnownBits Known = DAG.computeKnownBits(N00);
+        if (IsRightShift)
+          Known = Known.reverseBits();
+        if (ShAmt->ule(Known.countMinLeadingZeros()) ||
+            ShAmt->ult(Known.countMinSignBits()))
```

Comment:
Do these knownbits check ever succeed? Think that we will have always set the respective flag if they are true.

Reply:
Are you saying that SelectionDAG/DAGCombiner is deriving those flags somewhere? It is unclear to me if that actually is done.
Or are you assuming that the input to SelectionDAG already has these flags (and that lowering/legalization always are setting the flags when introducing shifts)?

Anyway, there is at least one special case here that isn't covered by the flags (although no idea how important it is). But imagine a right shift where the five least significant bits are known to be either one or zero. If shifting 4 steps the shift isn't exact, but we still know that the zero:edness doesn't change, as the shifted out bits are equal to at least one of the remaining bits.

But I think that in general poison generating flags mainly should be used when needed to imply something that isn't known otherwise (such as UB on signed overflow). When used for a condition that can be derived by value tracking those flags are just caching some information. But there aren't really any rules that say that it isn't OK to drop the flags, or that a transform must derive that information, right(?).
We are for example dropping poison generating flags when moving freeze instruction through the DAG. So they may disappear during DAG combine, while I think that the transform here still would be legal. Therefore it seems weird to base the transform on the existence of the flags.

Reply:
The test case here for example https://github.com/llvm/llvm-project/pull/88801/files/f603024d54df8b87428e9eb3ddb6b44856c28b5a#diff-d54fdf520f394753a79ba9e32d77d40bc4ffd7f1887c2710f5a067cab847e946 , is resulting in an optimized DAG like this:
```
Optimized type-legalized selection DAG: %bb.0 'i16_cmpz:entry'
SelectionDAG has 13 nodes:
  t0: ch,glue = EntryToken
          t2: i32,ch = CopyFromReg t0, Register:i32 %0
        t40: i32 = and t2, Constant:i32<65024>
      t34: i32 = srl t40, Constant:i32<9>
    t41: ch = br_cc t0, setne:ch, t34, Constant:i32<0>, BasicBlock:ch<if.end 0x5559bb945028>
  t14: ch = br t41, BasicBlock:ch<if.then 0x5559bb944f28>
```
where the SRL isn't marked as exact even if all shifted out bits are zero.

Reply:
what analysis do we have that lets us know consecutive trailing bits w.o confirming they are either zeros/non-zeros. If they are non-zeros we can just simplify the expression to false.

You are right, we don't have logic to add flags in `visit{SHL,SAR,SHR}`. Ill make patches for that as it seems like a better approach to be able to rely on those flags rather than essentially implementing the logic to check for them here.

Reply:
Could we add the flags during the i16->i32 promotion?

Reply:
IMHO it is a waste of time to derive these things, just in case someone wants to use them in the future. We do run DAG combiner several times, so for example adding some logic in visitSHL to try to derive the flags, potentially failing each time, seems like a real waste if there is no DAG combine that is interested in knowing that information. 

And I frankly don't know if poison generating flags is the solution for caching value tracking results. We could just as well wanna cache that we failed to derive that the shift is `exact`, or that a particular shift is known to not be `exact`. But not sure how to deal with that in a nice way.

I think it is much better to find a way were we can compute these things lazily when a fold is asking for the information, rather than expecting that the flags always are set when possible. OTOH, when some optimization is interested to know these kinds of properties, and derives information via value tracking, then we might wanna cache that information somehow to avoid having to compute it again (mutate the DAG and add the flags).

In this particular case we will fold away the shift if the flags are set. So there is no shift left to add the flags to after doing the fold.

Regardless, if the community decides that flags always should be derived up-front instead of being lazily computed based on need, then implementing that strategy is out-of-scope for this pull request. I think we really need a much better framework to automatically add flags when creating new nodes, as well as much better documentation regarding which posion generating flags that are mandatory to set on which DAG nodes (including some kind of verifier support) if we want to go that way.

Reply:
The middle end optimizations should have done the hard parts. We have to repeat a lot of optimizations in the DAG, where everything is slower and will do a worse job so ideally there would be a specific reason to do something.

I think overall it would be better to try to maintain information we already had, rather than re-infer it. For the motivating promotion case, SHL looks easy and missed today: https://alive2.llvm.org/ce/z/asYzPY

The right shift cases are trickier since you need to inspect the input, which we probably shouldn't do eagerly